tortoise/organizer/organizer.go

```diff
@@ -0,0 +1,122 @@
+package organizer
+
+import (
+	"context"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+)
+
+// LayersIterator is used to traverse organized layers.
+// For siplicity LayersIterator iterates over all layers unconditionally.
+type LayersIterator func(types.LayerID)
+
+// Opt is for configuring Organizer.
+type Opt func(*Organizer)
+
+// WithLogger configures logger for Organizer.
+func WithLogger(logger log.Log) Opt {
+	return func(o *Organizer) {
+		o.logger = logger
+	}
+}
+
+// WithLastLayer layer configures last layer that tortoise knows about.
+func WithLastLayer(lid types.LayerID) Opt {
+	return func(o *Organizer) {
+		o.submitted = lid
+		o.received = lid
+	}
+}
+
+// WithWindowSize configures size of the window.
+// By default window of the size 1000 is used.
+func WithWindowSize(size int) Opt {
+	return func(o *Organizer) {
+		o.window = make([]types.LayerID, size)
+	}
+}
+
+// New creates Organizer instance.
+func New(opts ...Opt) *Organizer {
+	o := &Organizer{
+		logger: log.NewNop(),
+		window: make([]types.LayerID, 1000), // ~4kb
+	}
+	for _, opt := range opts {
+		opt(o)
+	}
+	return o
+}
+
+// Organizer is a decorator for tortoise that keeps layers in order within the window
+// and sends them to tortoise only when the window has no gaps.
+type Organizer struct {
+	logger log.Log
+
+	// last submitted layer
+	submitted types.LayerID
+	// last received layer that is not submitted
+	received types.LayerID
+	// all layers between submitted and received
+	// window is non-empty only if there are gaps after submitted
+	window []types.LayerID
+}
+
+// Iterate allows to iterate over all ordered layers that are currently within the window including new layer.
+func (o *Organizer) Iterate(ctx context.Context, lid types.LayerID, f LayersIterator) {
+	if !lid.After(o.submitted) {
+		return
+	}
+	if lid.After(o.submitted.Add(uint32(len(o.window)))) {
```

Comment:
this may be overkill. in hare there are two params to make sure there are limited layers of delay by hare.
`LimitIterations` and `LimitConcurrent`. one can do a startup check to make sure the window size we use here make sense. i think even a buffer of 100 layers in the mainnet is too many.

Reply:
this is also to help with https://github.com/spacemeshos/go-spacemesh/issues/2782 , verified layer may progress much slower compared with synced layer

in general i think that code needs to be robust without worrying about the specifics of the hare or other parts. if small window is sufficient it can be defined at the startup with option. but for this module to be correct it needs to handle window overflow, i thought about three options in such case:
- leak memory
- panic
- apply layers with a gap

Reply:
my understanding of the code, wrt https://github.com/spacemeshos/go-spacemesh/issues/2782, is that we picked a random base block greedily. we did not try to find a base block that has the least diff (most similar opinions) from our own local opinions, which cause this to happen more than it should.

i am not comparing verified layer with synced/processed layer. but with the hare params, i cannot imagine why we need 1000 buckets for delayed verified layers.

Reply:
as i mentioned this is configurable, it just costs nothing to use 1000 that's why i set it as default.

regarding https://github.com/spacemeshos/go-spacemesh/issues/2782, as i roughly understand it works as:
1. a node with a gap in history created block A in layer 12 with base block B from layer 10
2. another node registered block A as good and used it as a base block for block C in layer 13. because block A misses opinions for block in layer 11, we add them as exceptions
3. every node will not mark block C as good block 

so there are several problems:
1. creating blocks with gaps in history is harmful, and it can either happen due to out-of-order hare or delay between synced and verified layers
2. base block selection needs adjusting

Reply:
> regarding #2782, as i roughly understand it works as:
> 
> 1. a node with a gap in history created block A in layer 12 with base block B from layer 10
> 2. another node registered block A as good and used it as a base block for block C in layer 13. because block A misses opinions for block in layer 11, we add them as exceptions
> 3. every node will not mark block C as good block

in my view, this still comes down to selecting base block with the least diff.
i.e. the node that produce block C should be able to find another base block that doesn't have gap in layer 11 as the minimal diff, instead of using block A since it's the first good block it encounters.

granted if all nodes are stuck, then the network is still left without a good base block.

Reply:
i agree regarding base blocks and i am going to implement the algorithm described in https://github.com/spacemeshos/research/issues/33#issuecomment-830448138 with some modifications.

but also we should not be even creating such blocks if we can, this is why the buffer is set to 1000 here, because it costs nothing but makes behavior in line with the correct behavior. do you disagree with this or with something else?

Reply:
i don't disagree with 1000. like you said it costs nothing. i am just not able to wrap my brains around how it could lag till 1000 once layer order is enforced. let's move ahead with what you currently have. thanks.